src/middlewared/middlewared/plugins/update.py

```diff
@@ -645,3 +656,28 @@ def destroy_upload_location(self):
         )
         if cp.returncode != 0:
             raise CallError(f'Could not destroy memory device: {cp.stderr}')
+
+    @private
+    def take_systemdataset_samba4_snapshot(self, new_version):
+        basename = self.middleware.call_sync('systemdataset.config')['basename']
+        if basename is None:
+            self.logger.warning('System dataset is not available, not taking snapshot')
+            return
+
+        dataset = f'{basename}/samba4'
+
+        proc = subprocess.run(['zfs', 'list', '-t', 'snapshot', '-H', '-o', 'name', '-s', 'name', '-d', '1', dataset],
+                              stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding='utf8', errors='ignore')
+        if proc.returncode != 0:
+            self.logger.warning('Unable to list dataset %s snapshots: %s', dataset, proc.stderr)
+            return
+
+        snapshots = [s.split('@')[1] for s in proc.stdout.strip().split()]
+        for snapshot in [s for s in snapshots if s.startswith('update-')][:-4]:
+            self.logger.info('Deleting dataset %s snapshot %s', dataset, snapshot)
+            subprocess.run(['zfs', 'destroy', f'{dataset}@{snapshot}'])
+
+        current_version = "-".join(self.middleware.call_sync("system.info")["version"].split("-")[1:])
+        new_version = "-".join(new_version.split("-")[1:])
```

Comment:
I am not seeing the usefulness of using new version here, perhaps just stick with the version its running?

Reply:
> I am not seeing the usefulness of using new version here, perhaps just stick with the version its running?

Yeah, let's just use the current version it's running at the time of snapshot.